File: test_materials.js
import * as WebIFC from "web-ifc";
import fs from "fs";

async function run() {
  const IfcAPI = new WebIFC.IfcAPI();
  IfcAPI.SetWasmPath("./node_modules/web-ifc/");
  await IfcAPI.Init();

  const fileData = fs.readFileSync("./01_BIMcollab_Example_ARC.ifc");

  const modelID = IfcAPI.OpenModel(new Uint8Array(fileData));

  const rels = IfcAPI.GetLineIDsWithType(modelID, WebIFC.IFCRELASSOCIATESMATERIAL);
  let matMap = {};
  for (let i = 0; i < rels.size(); i++) {
    const relID = rels.get(i);
    const rel = IfcAPI.GetLine(modelID, relID);
    
    let matName = "Unknown";
    try {
        const mat = IfcAPI.GetLine(modelID, rel.RelatingMaterial.value);
        if (mat.Name && mat.Name.value) {
            matName = mat.Name.value;
        } else if (mat.MaterialLayers) {
            const layer = IfcAPI.GetLine(modelID, mat.MaterialLayers[0].value);
            const m = IfcAPI.GetLine(modelID, layer.Material.value);
            if (m && m.Name) matName = m.Name.value;
        } else if (mat.Materials) {
            const m = IfcAPI.GetLine(modelID, mat.Materials[0].value);
            if (m && m.Name) matName = m.Name.value;
        } else if (mat.ForLayerSet) {
            const layerSet = IfcAPI.GetLine(modelID, mat.ForLayerSet.value);
            if (layerSet && layerSet.MaterialLayers) {
                const layer = IfcAPI.GetLine(modelID, layerSet.MaterialLayers[0].value);
                const m = IfcAPI.GetLine(modelID, layer.Material.value);
                if (m && m.Name) matName = m.Name.value;
            }
        }
    } catch (e) {
    }

    if (rel.RelatedObjects) {
        for(let j=0; j<rel.RelatedObjects.length; j++) {
            const objId = rel.RelatedObjects[j].value;
            matMap[objId] = matName;
        }
    }
  }

  console.log("Material Map entries:", Object.keys(matMap).length);
  const keys = Object.keys(matMap);
  if(keys.length > 0) {
      console.log("Example 1:", keys[0], matMap[keys[0]]);
      console.log("Example 2:", keys[1], matMap[keys[1]]);
  } else {
      console.log("NO MATERIALS FOUND!");
  }

  IfcAPI.CloseModel(modelID);
}
run().catch(console.error);

File: test_materials.py
import ifcopenshell
import sys

def run():
    ifc_file = "01_BIMcollab_Example_ARC.ifc"
    f = ifcopenshell.open(ifc_file)
    
    mat_count = 0
    mat_names = set()
    
    for rel in f.by_type("IfcRelAssociatesMaterial"):
        mat_count += 1
        mat = rel.RelatingMaterial
        name = "Unknown"
        
        if mat.is_a("IfcMaterial"):
            name = mat.Name
        elif mat.is_a("IfcMaterialLayerSet"):
            if mat.MaterialLayers:
                name = mat.MaterialLayers[0].Material.Name
        elif mat.is_a("IfcMaterialLayerSetUsage"):
            if mat.ForLayerSet.MaterialLayers:
                name = mat.ForLayerSet.MaterialLayers[0].Material.Name
        elif mat.is_a("IfcMaterialList"):
            if mat.Materials:
                name = mat.Materials[0].Name
        elif mat.is_a("IfcMaterialProfileSet"):
            if mat.MaterialProfiles:
                name = mat.MaterialProfiles[0].Material.Name
        elif mat.is_a("IfcMaterialProfileSetUsage"):
            if mat.ForProfileSet.MaterialProfiles:
                name = mat.ForProfileSet.MaterialProfiles[0].Material.Name
        
        mat_names.add(name)
        
        if mat_count < 10:
            print(f"Rel: {rel.id()}, MatType: {mat.is_a()}, FoundName: {name}")

    print(f"Total Material Relations: {mat_count}")
    print(f"Unique Materials: {mat_names}")

if __name__ == "__main__":
    run()
